docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,500 @@
+---
+
+# How to migrate from Newtonsoft.Json to System.Text.Json
+
+This article shows how to migrate from [Newtonsoft.Json](https://www.newtonsoft.com/json) to [System.Text.Json](system-text-json-overview.md).
+
+ `System.Text.Json` focuses primarily on performance, security, and standards compliance. It has some key differences in default behavior and doesn't aim to have feature parity with `Newtonsoft.Json`. For some scenarios, `System.Text.Json` has no built-in functionality, but there are recommended workarounds. For other scenarios, workarounds are impractical. If your application depends on a missing feature, consider [filing an issue](https://github.com/dotnet/runtime/issues/new) to find out if support for your scenario can be added.
+
+<!-- For information about which features might be added in future releases, see the [Roadmap](https://github.com/dotnet/runtime/tree/master/src/libraries/System.Text.Json/roadmap/README.md). [Restore this when the roadmap is updated.]-->
+
+Most of this article is about how to use the <xref:System.Text.Json.JsonSerializer> API, but it also includes guidance on how to use the <xref:System.Text.Json.JsonDocument> (which represents the Document Object Model or DOM), <xref:System.Text.Json.Utf8JsonReader>, and <xref:System.Text.Json.Utf8JsonWriter> types. The article is organized into sections in the following order:
+
+* [Differences in **default** JsonSerializer behavior compared to Newtonsoft.Json](#differences-in-default-jsonserializer-behavior-compared-to-newtonsoftjson)
+* [Scenarios using JsonSerializer that require workarounds](#scenarios-using-jsonserializer-that-require-workarounds)
+* [Scenarios that JsonSerializer currently doesn't support](#scenarios-that-jsonserializer-currently-doesnt-support)
+* [JsonDocument and JsonElement compared to JToken (like JObject, JArray)](#jsondocument-and-jsonelement-compared-to-jtoken-like-jobject-jarray)
+* [Utf8JsonReader](#utf8jsonreader)
+* [Utf8JsonWriter](#utf8jsonwriter)
+
+## Differences in default JsonSerializer behavior compared to Newtonsoft.Json
+
+`System.Text.Json` is strict by default and avoids any guessing or interpretation on the caller's behalf, emphasizing deterministic behavior. The library is intentionally designed this way for performance and security. `Newtonsoft.Json` is flexible by default.
+
+### Case-insensitive deserialization 
+
+During deserialization, `Newtonsoft.Json` does case-insensitive property name matching by default. The `System.Text.Json` default is case-sensitive, which gives better performance since it's doing an exact match. For information about how to do case-insensitive matching, see [Case-insensitive property matching](system-text-json-how-to.md#case-insensitive-property-matching).
+
+If you're using `System.Text.Json` indirectly by using ASP.NET Core, you don't need to do anything to get behavior like `Newtonsoft.Json`. ASP.NET Core specifies the case-insensitive setting when it uses `System.Text.Json`.
```

Comment:
Can a developer select case-sensitive comparisons in an ASP.NET Core application? Or would that break dependencies in ASP.NET Core?

Reply:
>Can a developer select case-sensitive comparisons in an ASP.NET Core application? Or would that break dependencies in ASP.NET Core?

You can select it, but whether it would break anything other than app code that depends on the Newtonsoft default I don't know.  Maybe @pranavkm  or @ahsonkhan  would know.

Reply:
> Or would that break dependencies in ASP.NET Core?

It shouldn't. As far as I know, it will just affect model binding semantics, so certain payloads won't bind to the API because of mismatch due to casing. If the developer wants that behavior, it is legitimate to do so. @pranavkm can correct me or expand on that, if needed :)

Reply:
Thanks @ahsonkhan 

Here's what raised the question in my mind: I've written a few small apps that use the GitHub GraphQL and REST APIs. Most are console apps, one is a web site. I've got unit tests on the types I've created that encapsulate the JsonElement types. This made me wonder if there would be different behavior in the unit tests, the console apps, and the web app. And, if so, what would be the best way for a developer to ensure consistent behavior. That would make it easier to trust the unit tests.

Reply:
> GitHub GraphQL and REST APIs

Is the JSON responses from those always camel cased, or is it sometime pascal case/mixed? Depending on your inputs and scenario, you either want to set the `PropertyNamingPolicy` or `PropertyNameCaseInsensitive` or both. Are you only ever reading or do you ever write JSON?

> I've got unit tests on the types I've created that encapsulate the JsonElement types

The case-insensitive vs case-sensitive property name comparison only applies to `JsonSerializer` and `JsonSerializerOptions`. `JsonElement` always does case-sensitive (exact) match, so the question doesn't apply there.

> This made me wonder if there would be different behavior in the unit tests, the console apps, and the web app.

I would have to see the details of the code to better answer this. That said, given the default `JsonSerializer` behavior differs from what aspnet uses as its defaults, you will see a difference between uses within console apps vs web apps. One way to make it consistent is to use the same options across the board (so either change the uses of `JsonSerializerOptions` in the console app to match what aspnet is doing, or vice versa).

These are the defaults settings of the `JsonSerializer` within aspnet:
https://github.com/dotnet/aspnetcore/blob/a6bc6ce23dad5a9d02596cf2e91e3c4f965c61bc/src/Mvc/Mvc.Core/src/JsonOptions.cs#L15-L29